File: w3/w3java/Basic/Point/Point.java
public class Point implements Cloneable {
    private double x;
    private double y;
    
    public Point(double x, double y) {
        this.x = x;
        this.y = y;
    }
    
    public Point() {
        this.x = 0.0;
        this.y = 0.0;
    }
    
    public double getX() {
        return this.x;
    }
    
    public double getY() {
        return this.y;
    }
    
    public void setX(double value) {
        this.x = value;
    }
    
    public void setY(double value) {
        this.y = value;
    }
    
    public double distance(Point other) {
        return Math.hypot(this.x-other.x, this.y-other.y);
    }
    
    public boolean equals(Point other) {
        return this.x == other.x && this.y == other.y;
    }
    
    public String toString() {
        return "(" + x + ", " + y + ")";
    }
    
    public Point clone() throws CloneNotSupportedException {
        return (Point) super.clone();
    }
}

File: w3/w3cpp/Car/Point.cpp
#include <iostream>
#include "Point.h"

Point::Point(double x, double y) {
    this->x = x;
    this->y = y;
}

Point::~Point() {
    
}

double Point::getX() const {
    return this->x;
}

double Point::getY() const {
    return this->y;
}

// void Point::setX(double value) {
//     this->x = value;
// }

// void  Point::setY(double value) {
//     this->y = value;
// }

double Point::distance(const Point& other) const {
    return hypot(this->x-other.x, this->y-other.y);
}

bool Point::operator==(const Point& other) const {
    if ( this->x == other.x && this->y == other.y ) {
        return true;
    }
    return false;
}

bool Point::operator!=(const Point& other) const {
    if ( this->x != other.x || this->y != other.y ) {
        return true;
    }
    return false;
}

std::ostream& operator<<(std::ostream& out, const Point& point) {
    out << '(' << point.getX() << ", " << point.getY() << ')';
    return out;
}
